crates/nu-cli/src/repl.rs

```diff
@@ -729,6 +715,44 @@ pub fn evaluate_repl(
     Ok(())
 }
 
+fn store_history_id_in_engine(engine_state: &mut EngineState, line_editor: &Reedline) {
+    let session_id = line_editor
+        .get_history_session_id()
+        .map(i64::from)
+        .unwrap_or(0);
+
+    engine_state.history_session_id = session_id;
+}
+
+fn update_line_editor_history(
+    engine_state: &mut EngineState,
+    history_path: &Path,
+    line_editor: Reedline,
+    history_session_id: Option<HistorySessionId>,
+) -> Result<Reedline, ErrReport> {
+    let config = engine_state.get_config();
+    let history: Box<dyn reedline::History> = match engine_state.config.history_file_format {
+        HistoryFileFormat::PlainText => Box::new(
+            FileBackedHistory::with_file(
+                config.max_history_size as usize,
+                history_path.to_path_buf(),
+            )
+            .into_diagnostic()?,
+        ),
+        HistoryFileFormat::Sqlite => {
+            Box::new(SqliteBackedHistory::with_file(history_path.to_path_buf()).into_diagnostic()?)
+        }
+    };
+    let line_editor = line_editor
+        .with_history_session_id(history_session_id)
+        .with_history_exclusion_prefix(Some(" ".into()))
+        .with_history(history);
+
+    store_history_id_in_engine(engine_state, &line_editor);
```

Comment:
> i'm just having trouble seeing where the changes are fixing the bug 🤔

Sorry I should've mentioned that in the description !

> i'm just having trouble seeing where the changes are fixing the bug 🤔

This is the only change that I did apart from the refactor.
When we were updating the line_editor's `history_session_id` that came from `history_isolation` we did not give it the `engine_state` : 
```rust
// Setup history_isolation aka "history per session"
let history_isolation = config.history_isolation;
let history_session_id = if history_isolation {
    Reedline::create_history_session_id()
} else {
    None
};
```

Reply:
> Sorry I should've mentioned that in the description !
>
no worries :wink: 

> This is the only change that I did apart from the refactor.
>
i'm confused about that...
i see this snippet of code in `crates/nu-cli/src/repl.rs` but you did not change anything about _isolation_ or _giving to the `engine_state`_ in the PR, did you? :confused:

Reply:
maybe it would be more direct if you could point me directly in the diff of this PR where the real meat occurs? :yum: 

i might be just completely blind and dumb here :laughing:

Reply:
> i see this snippet of code in crates/nu-cli/src/repl.rs but you did not change anything about isolation or giving to the engine_state in the PR, did you? 😕

I did not change anything about _isolation_ but I did change something about _giving to the `engine_state`_.
The main change is in this diff : 
https://github.com/nushell/nushell/pull/9616/files/bc4b375f7b9a43256a321206d88b881d1edff831#diff-665392adc0e17b5c199c59ae562f98f0673e18effdb813dc3578a1e55e191e42L130-L145

As you can see here we did not previously give the new `history_session_id` to the `engine_state`.
Which now we do in the new function that I introduced (`update_line_editor_history`) : 
https://github.com/nushell/nushell/pull/9616/files/bc4b375f7b9a43256a321206d88b881d1edff831#diff-665392adc0e17b5c199c59ae562f98f0673e18effdb813dc3578a1e55e191e42R751

Reply:
ooooooooh, thank you so much :pray: 

i just completely did not notice this new line
```rust
    store_history_id_in_engine(engine_state, &line_editor);
```

Reply:
@YassineHaouzane 
could you update the description of the PR for future reviewers? :innocent: 
that'd be just perfect :relieved: 

something that mentions that the PR is mainly refactoring these into functions AND adding one call to update the `engine_state` :wink:

Reply:
(and you commented on your changes just on the line of interest... my bad, i was confused :scream: :laughing:)

Reply:
>could you update the description of the PR for future reviewers? 😇
Done 👍

Reply:
thanks a lot :blush: